agent/config/file_watcher.go

```diff
@@ -0,0 +1,221 @@
+package config
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/hashicorp/go-hclog"
+)
+
+const timeoutDuration = 200 * time.Millisecond
+
+type FileWatcher struct {
+	watcher          *fsnotify.Watcher
+	configFiles      map[string]*watchedFile
+	handleFunc       func(event *WatcherEvent)
+	logger           hclog.Logger
+	reconcileTimeout time.Duration
+	cancel           context.CancelFunc
+	done             chan interface{}
+}
+
+type watchedFile struct {
+	id uint64
+}
+
+type WatcherEvent struct {
+	Filename string
+}
+
+func NewFileWatcher(handleFunc func(event *WatcherEvent), configFiles []string, logger hclog.Logger) (*FileWatcher, error) {
+	ws, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	cfgFiles := make(map[string]*watchedFile)
+
+	w := &FileWatcher{watcher: ws,
+		configFiles:      cfgFiles,
+		handleFunc:       handleFunc,
+		reconcileTimeout: timeoutDuration,
+		done:             make(chan interface{}),
+	}
+	w.logger = logger.Named("file-watcher")
+	for _, f := range configFiles {
+		err = w.add(f)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return w, nil
+}
+
+func (w *FileWatcher) Start(ctx context.Context) {
+	cancelCtx, cancel := context.WithCancel(ctx)
+	w.cancel = cancel
+	go w.watch(cancelCtx)
+}
+
+func (w *FileWatcher) add(filename string) error {
+	if isSymLink(filename) {
+		return fmt.Errorf("symbolic link are not supported %s", filename)
+	}
+	filename = filepath.Clean(filename)
+	w.logger.Trace("adding file", "file", filename)
+	if err := w.watcher.Add(filename); err != nil {
+		return err
+	}
+	iNode, err := w.getFileId(filename)
+	if err != nil {
+		return err
+	}
+	w.configFiles[filename] = &watchedFile{id: iNode}
+	return nil
+}
+
+func isSymLink(filename string) bool {
+	fi, err := os.Lstat(filename)
+	if err != nil {
+		return false
+	}
+	if fi.Mode()&os.ModeSymlink != 0 {
+		return true
+	}
+	return false
+}
+
+func (w *FileWatcher) Stop() error {
+	w.cancel()
+	<-w.done
+	return w.watcher.Close()
+}
+
+func (w *FileWatcher) watch(ctx context.Context) {
+	ticker := time.NewTicker(w.reconcileTimeout)
+	defer ticker.Stop()
+	for {
+		select {
+		case event, ok := <-w.watcher.Events:
+			if !ok {
+				w.logger.Error("watcher event channel is closed")
+				return
+			}
+			w.logger.Debug("received watcher event", "event", event)
+			if err := w.handleEvent(event); err != nil {
+				w.logger.Error("error handling watcher event", "error", err, "event", event)
+			}
+		case _, ok := <-w.watcher.Errors:
+			if !ok {
+				w.logger.Error("watcher error channel is closed")
+				return
+			}
+		case <-ticker.C:
+			w.reconcile()
+		case <-ctx.Done():
+			close(w.done)
```

Comment:
Should this actually be in a defer on line 99 instead? (top of the function). You have several ways to exit this method, but only one of them closes the done channel.